File: epi_judge_java_solutions/epi/SortAlmostSortedArray.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.PriorityQueue;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class SortAlmostSortedArray {

  public static List<Integer>
  sortApproximatelySortedData(Iterator<Integer> sequence, int k) {

    PriorityQueue<Integer> minHeap = new PriorityQueue<>();
    // Adds the first k elements into minHeap. Stop if there are fewer than k
    // elements.
    for (int i = 0; i < k && sequence.hasNext(); ++i) {
      minHeap.add(sequence.next());
    }

    List<Integer> result = new ArrayList<>();
    // For every new element, add it to minHeap and extract the smallest.
    while (sequence.hasNext()) {
      minHeap.add(sequence.next());
      Integer smallest = minHeap.remove();
      result.add(smallest);
    }

    // sequence is exhausted, iteratively extracts the remaining elements.
    result.addAll(Stream.generate(minHeap::remove)
                      .limit(minHeap.size())
                      .collect(Collectors.toList()));
    return result;
  }

  @EpiTest(testDataFile = "sort_almost_sorted_array.tsv")
  public static List<Integer>
  sortApproximatelySortedDataWrapper(List<Integer> sequence, int k) {
    return sortApproximatelySortedData(sequence.iterator(), k);
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "SortAlmostSortedArray.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}

File: epi_judge_cpp_solutions/sort_almost_sorted_array.cc
#include <functional>
#include <queue>
#include <string>
#include <vector>

#include "test_framework/generic_test.h"

using std::greater;
using std::priority_queue;
using std::string;
using std::vector;

vector<int> SortApproximatelySortedData(
    vector<int>::const_iterator sequence_begin,
    const vector<int>::const_iterator& sequence_end, int k) {
  priority_queue<int, vector<int>, greater<>> min_heap;
  // Adds the first k elements into min_heap. Stop if there are fewer than k
  // elements.
  for (int i = 0; i < k && sequence_begin != sequence_end; ++i) {
    min_heap.push(*sequence_begin++);
  }

  vector<int> result;
  // For every new element, add it to min_heap and extract the smallest.
  while (sequence_begin != sequence_end) {
    min_heap.push(*sequence_begin++);
    result.push_back(min_heap.top());
    min_heap.pop();
  }

  // sequence is exhausted, iteratively extracts the remaining elements.
  while (!empty(min_heap)) {
    result.push_back(min_heap.top());
    min_heap.pop();
  }

  return result;
}

vector<int> SortApproximatelySortedDataWrapper(const vector<int>& sequence,
                                               int k) {
  return SortApproximatelySortedData(cbegin(sequence), cend(sequence), k);
}

// clang-format off


int main(int argc, char* argv[]) {
  std::vector<std::string> args {argv + 1, argv + argc};
  std::vector<std::string> param_names {"sequence", "k"};
  return GenericTestMain(args, "sort_almost_sorted_array.cc", "sort_almost_sorted_array.tsv", &SortApproximatelySortedDataWrapper,
                         DefaultComparator{}, param_names);
}
// clang-format on
